lib/vm/src/instance.rs

```diff
@@ -70,46 +71,56 @@ where
                 }),
             },
         };
-        let instance = module.instantiate(&import_obj).context(WasmerErr {})?;
+        let wasmer_instance = module.instantiate(&import_obj).context(WasmerErr {})?;
+        Ok(Instance::from_wasmer(wasmer_instance, deps))
+    }
+
+    pub fn from_wasmer(wasmer_instance: wasmer_runtime_core::Instance, deps: Extern<S, A>) -> Self {
         let res = Instance {
-            instance,
-            api,
-            storage: PhantomData::<S> {},
+            wasmer_instance: wasmer_instance,
+            api: deps.api,
+            type_storage: PhantomData::<S> {},
         };
         res.leave_storage(Some(deps.storage));
-        Ok(res)
+        res
+    }
+
+    /// Takes ownership of instance and decomposes it into its components.
+    /// The components we want to preserve are returned, the rest is dropped.
+    pub fn recycle(instance: Self) -> (wasmer_runtime_core::Instance, A) {
```

Comment:
What about adding `take_storage` in here, which is tied to the wasmer_instance context.
So it fully decomposes all the info we put in.

Reply:
I liked the idea. However, when thinking about it more I figured that this would make `Instance` construction and `Instance::recycle` asymmetric. This is because no representation of storage is owned by the instance. I think when we manually group storage and instance using `leave_storage`, we should also manually ungroup then later.

Reply:
The usage I see (from the high level) is:

`Instance::from_module(&module, deps)` or `Instance::from_code(&wasm, deps)` to construct it. This takes `deps: Extern<S, A>`.

When freeing, or recycling the instance, I would assume the inverse of something like:

`let {wasmer_instance, deps} = instance::recycle();`

which should return a comprable `deps` as when we constructed it.

On another note, I wonder if we can make `leave_storage` and `take_storage` private to just the module. On the public side, we could create a new instance with one of the `from_xxx()` methods, `recycle` it, or do some queries via `instance.with_storage(&|store| { ... })`.

This can be another cleanup PR, but I think will simplify the external API and the safety guarantees. (I really like the `recycle` idea)

Reply:
> The usage I see (from the high level) is

oh yeah, right, this way we have symmetry.

> On another note, I wonder if we can make leave_storage and take_storage private to just the module.

Jap, that works!